app/src/test/java/io/hstream/testing/ScaleTest.java

```diff
@@ -0,0 +1,288 @@
+    Set<RecordId> set1 = new HashSet<>();
+    CountDownLatch countDown1 = new CountDownLatch(total * total);
+    Consumer consumer =
+        hStreamClient
+            .newConsumer()
+            .subscription(subscription)
+            .rawRecordReceiver(
+                (recs, recv) -> {
+                  recv.ack();
+                  if (set1.add(recs.getRecordId())) {
+                    countDown1.countDown();
+                  }
+                })
+            .build();
+
+    consumer.startAsync().awaitRunning();
+    Assertions.assertTrue(countDown1.await(90, TimeUnit.SECONDS));
+    consumer.stopAsync().awaitTerminated();
+
+    Assertions.assertEquals(set0, set1);
+    Assertions.assertEquals(total * total, recordIds0.size());
+  }
+
+  @Test
+  @Timeout(120)
+  void testLargeNumConsumer() throws Exception {
+    final String stream = randStream(hStreamClient);
+    final String subscription = randSubscription(hStreamClient, stream);
+    final int total = 64;
+
+    Producer producer = hStreamClient.newProducer().stream(stream).build();
+    List<RecordId> recordIds0 = new ArrayList<>();
+    for (int i = 0; i < total; ++i) {
+      recordIds0.add(
+          producer.write(UUID.randomUUID().toString().getBytes(StandardCharsets.UTF_8)).join());
+    }
+
+    AtomicReference<Throwable> err = new AtomicReference<>();
+    err.set(null);
+    List<Thread> threads = new ArrayList<>();
+    for (int i = 0; i < total; ++i) {
+      Thread thread =
+          new Thread(
+              () -> {
+                Set<RecordId> recordIds1 = new HashSet<>();
+                CountDownLatch countDown = new CountDownLatch(total);
+                Consumer consumer =
+                    hStreamClient
+                        .newConsumer()
+                        .subscription(subscription)
+                        .rawRecordReceiver(
+                            (recs, recv) -> {
+                              if (recordIds1.add(recs.getRecordId())) {
+                                countDown.countDown();
+                              }
+                            })
+                        .build();
+                consumer.startAsync().awaitRunning();
+                try {
+                  Assertions.assertTrue(countDown.await(60, TimeUnit.SECONDS));
+                } catch (Throwable e) {
+                  if (err.get() != null) {
+                    err.set(e);
+                  }
+                }
+                consumer.stopAsync().awaitTerminated();
+                try {
+                  Assertions.assertEquals(new HashSet<>(recordIds0), recordIds1);
+                } catch (Throwable e) {
+                  if (err.get() != null) {
+                    err.set(e);
+                  }
+                }
+              });
+      thread.start();
+      threads.add(thread);
+    }
+    for (var x : threads) {
+      x.join();
+    }
+    Assertions.assertNull(err.get());
```

Comment:
It is so weird here... I recommend you re-checking what you want to test again.

Reply:
thx c50bcf3